scarb/src/compiler/plugin/proc_macro/compilation.rs

```diff
@@ -117,17 +196,34 @@ impl From<OutputFormat> for CargoOutputFormat {
     }
 }
 
-impl From<CargoCommand> for Command {
-    fn from(args: CargoCommand) -> Self {
+impl<'c> From<CargoCommand<'c>> for Command {
+    fn from(args: CargoCommand<'c>) -> Self {
         let mut cmd = Command::new(Tool::Cargo.path());
         cmd.current_dir(args.current_dir);
         match args.action {
             CargoAction::Fetch => cmd.arg("fetch"),
             CargoAction::Build => cmd.arg("build"),
             CargoAction::Check => cmd.arg("check"),
+            CargoAction::Package(_) => cmd.arg("package"),
         };
+        if args.config.offline() {
+            cmd.arg("--offline");
+        }
         match args.action {
             CargoAction::Fetch => (),
+            CargoAction::Package(ref opts) => {
+                cmd.arg("--target-dir");
+                cmd.arg(args.target_dir);
+                if !opts.check_metadata {
+                    cmd.arg("--no-metadata");
+                }
+                if !opts.verify {
+                    cmd.arg("--no-verify");
+                }
```

Comment:
This actually stops working if you set verify to false, as you copy from target dir in 
```
            path: CARGO_MANIFEST_FILE_NAME.into(),
            contents: ArchiveFileContents::OnDisk(
                pkg.target_path(ws.config())
                    .into_child("package")
                    .into_child(crate_archive_basename)
                    .into_child(CARGO_MANIFEST_FILE_NAME)
                    .path_unchecked()
                    .to_path_buf(),
```
While `pkg.target_path(ws.config()).into_child("package").into_child(crate_archive_basename)` is only unpacked during Cargo verification - otherwise there will only be `crate_archive_basename.crate` file in `pkg.target_path(ws.config()).into_child("package")`.

Reply:
Thanks, great catch!

The obvious solution would be to remove this and always verify on Cargo side. 
But, what `verify = true` essentially does in Cargo is unpack the `.crate` file and then run a check:
> 3. Extract the .crate file and build it to verify it can build.
> This will rebuild your package from scratch to ensure that it can be built from a pristine state. 
> The --no-verify flag can be used to skip this step.

So, a more accurate solution imo, would be to manually unpack `.crate`, while still allowing to skip verification.

Unfortunately, there's no something like `cargo unpack`: https://github.com/rust-lang/cargo/issues/3577
Still, there's no need to reinvent the wheel - we can just reuse this segment from Cargo to handle the extraction:
https://github.com/rust-lang/cargo/blob/a4600184b8d6619ed0b5a0a19946dbbe97e1d739/src/cargo/ops/cargo_package.rs#L1110

What do you think?

Reply:
Discussed offline, addressed in #1643 (so it's easier to review 🙂)